File: codewars/4kyu/Hamming Numbers/main.go
package kata

func min(a, b uint) uint {
	if a < b {
		return a
	}

	return b
}

func Hammer(n int) uint {
	h := make([]uint, n)
	h[0] = 1

	var (
		nextTwo   uint = 2
		nextThree uint = 3
		nextFive  uint = 5
	)

	i, j, k := 0, 0, 0
	for m := 1; m < len(h); m++ {
		h[m] = min(nextTwo, min(nextThree, nextFive))
		if h[m] == nextTwo {
			i++
			nextTwo = 2 * h[i]
		}
		if h[m] == nextThree {
			j++
			nextThree = 3 * h[j]
		}
		if h[m] == nextFive {
			k++
			nextFive = 5 * h[k]
		}
	}

	return h[n-1]
}

File: codewars/4kyu/Hamming Numbers/main.rs
fn hamming(n: usize) -> u64 {
    let mut h = vec![0u64; n];
    h[0] = 1;

    let (mut next_two, mut next_three, mut next_five) = (2u64, 3u64, 5u64);
    let (mut i, mut j, mut k) = (0, 0, 0);

    for m in 1..n {
        let next_hamming = next_two.min(next_three.min(next_five));
        h[m] = next_hamming;

        if next_hamming == next_two {
            i += 1;
            next_two = 2 * h[i];
        }
        if next_hamming == next_three {
            j += 1;
            next_three = 3 * h[j];
        }
        if next_hamming == next_five {
            k += 1;
            next_five = 5 * h[k];
        }
    }

    h[n - 1]
}
